### DISCUSSION TITLE
自行编译paddlespeech-ctcdecoders的windows版本

### DISCUSSION THREAD

**Discussion by Doubledongli** [PaddlePaddle/PaddleSpeech#2148] [Show and tell]
> 朋友们，paddlespeech-ctcdecoders目前只有linux和macos版本，为了方便windows上门调试和学习，我自行编译了一个windows版本，欢迎指正。
### 1、安装基础环境
**（1）gcc**
![image](https://user-images.githubusercontent.com/20540661/178697507-9c080842-c383-4297-9d4f-00cd9921619a.png)
**（2）cmake**
![image](https://user-images.githubusercontent.com/20540661/178697702-ba2d3f33-4ff4-4897-b6d0-14d13666ce1a.png)
**（3）make**
![image](https://user-images.githubusercontent.com/20540661/178697971-e042f4e6-5578-4f68-a077-da55ae215a62.png)
**（4）vs**
![image](https://user-images.githubusercontent.com/20540661/178698474-3b6e63ee-f9d2-4d83-bd59-bbe7c0a0b130.png)
### 2、执行编译
```bash
cd PaddleSpeech\third_party\ctc_decoders
python -c++ -python decoders.i
```
执行完以后，会出现decoders_warp.cxx和paddlespeech_ctcdecoders.py，这一步一般不会出现问题，但是你编译后面的会报一大堆错误，所以我选择了另一种编译方式，也就是官网的setup.sh的编译方式。

**按下面方式编译**
```bash
git clone https://github.com/kpu/kenlm.git
wget http://www.openfst.org/twiki/pub/FST/FstDownload/openfst-1.6.3.tar.gz --no-check-certificate
tar -xzvf openfst-1.6.3.tar.gz
```
上面解压你可以选择git-bash窗口去解压，这个比较快
```bash
git clone https://github.com/progschj/ThreadPool.git
python3 setup.py install --num_processes 4
```
参考的文件内容其实如下：
```bash
#!/usr/bin/env bash

if [ ! -d kenlm ]; then
    git clone https://github.com/kpu/kenlm.git
    cd kenlm/
    git checkout df2d717e95183f79a90b2fa6e4307083a351ca6a
    cd ..
    echo -e "\n"
fi

if [ ! -d openfst-1.6.3 ]; then
    echo "Download and extract openfst ..."
    wget http://www.openfst.org/twiki/pub/FST/FstDownload/openfst-1.6.3.tar.gz --no-check-certificate
    tar -xzvf openfst-1.6.3.tar.gz
    echo -e "\n"
fi

if [ ! -d ThreadPool ]; then
    git clone https://github.com/progschj/ThreadPool.git
    echo -e "\n"
fi

echo "Install decoders ..."
python3 setup.py install --num_processes 4
```
### 3、编译日志
![image](https://user-images.githubusercontent.com/20540661/178702212-e440b2ca-7ff5-424b-83bf-bea34f02fb21.png)
![image](https://user-images.githubusercontent.com/20540661/178702374-ef1adeac-3a38-4c88-b7af-395580e13331.png)
### 4、发布whl包
![image](https://user-images.githubusercontent.com/20540661/178702659-67d06953-4e3b-4fad-b274-9b0617b39064.png)
### 测试
![image](https://user-images.githubusercontent.com/20540661/178702819-230e3e61-9dde-4526-9de9-3c1707494d7b.png)
**pypi包的获取**
[https://pypi.org/project/paddlespeech-ldd-ctcdecoders/](url)

### 5、总结

- 1、这个流程看起来很简单，但是中间我踩了无数的坑，具体的坑抽时间一个个写上去，如果你的流程中间有问题欢迎留言
- 2、目前包已经打好可以导入，但是是否能正常使用具体每个函数和功能还无从知晓，后期需要测试后再看，有问题欢迎交流
### 6、踩坑记录

1、setup.py文件里面下面的地方要注释掉，这个是用于linux或者是macos环境的，会报错一堆
![image](https://user-images.githubusercontent.com/20540661/178703501-04a3b5d9-86a6-442d-8671-8b554d07881b.png)
2、中间会有各种缺库文件的错误，暂时无法一一列举，后面更新

**Comment by lijielijie** [comment]
包怎么没了？

**Reply by yt605155624** [reply]
> 包怎么没了？

You can refer to this [third_party](https://github.com/PaddlePaddle/PaddleSpeech/tree/develop/third_party) to install it by yourself

**Comment by LFYG** [comment]
在编译一份py38的吧

**Comment by Circleee7** [comment]
编译安装了python import不了哇
![image](https://github.com/user-attachments/assets/d9e84307-7e00-4e43-b892-8167551c47c1)

**Reply by Circleee7** [reply]
而且安装的时候说找不到paddlespeech-ctcdecoders.py文件
